external/js/cothority/src/network/connection.ts

```diff
@@ -242,3 +250,234 @@ export class LeaderConnection extends WebSocketConnection {
+    // This can be set from the outside to enforce a number of parallel requests. If it
+    // is > 0, it will be enforced.
+    static parallel: number = 0;
+    // the threshold for the ratio of response_node / response_fastest where the node is
+    // put in the 'reserve'.
+    static slowThreshold: number = 10;
+    private readonly addresses: string[];
+    // to the slowest node.
+    private readonly active: Node[] = [];
+    // is the one that was not used for the longest time.
+    private readonly reserve: Node[];
+    // Pool of available connections that are not in the active connections. The first node
+
+    // Current number of nodes in the active queue.
+    private _parallel: number;
+    // Holds the fastest connections for 'parallel' nodes. They are sorted from the fastest
+
+    constructor(r: Roster, parallel: number = 3) {
+        this.addresses = r.list.map((conode) => conode.getWebSocketAddress());
+        shuffle(this.addresses);
+        // Initialize the pool of connections
+        this.reserve = this.addresses.map((addr) => new Node(addr));
+        this.parallel = parallel;
+    }
+
+    /**
+     * Creates a new NodeList for one message.
+     * @param service
+     */
+    newList(service: string): NodeList {
+        if (Nodes.parallel > 0) {
+            this.parallel = Nodes.parallel;
+        }
+        return new NodeList(this, service, this.active.slice(), this.reserve.slice());
+    }
+
+    /**
+     * Marks the node of the given address as having an error. The error must be
+     * a websocket-error 1001 or higher. An error in the request itself (refused
+     * transaction) should be treated as a passing node.
+     * @param address
+     */
+    gotError(address: string) {
+        this.replaceActive(this.index(address));
+    }
+
+    /**
+     * Marks the node with the given address as having successfully treated the
+     * message. It will re-order the nodes to reflect order of arrival. If the
+     * node is more than 10x slower than the fastest node, it will be replaced
+     * with a node from the reserve queue.
+     * @param address node with successful return
+     * @param rang order of arrival
+     * @param ratio delay in answer between first reply and this reply
+     */
+    done(address: string, rang: number, ratio: number) {
+        const index = this.index(address);
+        if (index >= 0) {
+            if (ratio >= Nodes.slowThreshold) {
+                this.replaceActive(index);
+            } else {
+                this.swapActive(index, rang);
+            }
+        }
+    }
+
+    /**
+     * Sets the timeout on all nodes.
+     * @param t
+     */
+    setTimeout(t: number) {
+        this.active.concat(this.reserve).forEach((n) => n.setTimeout(t));
+    }
+
+    /**
+     * Replaces the given node from the active queue with the first node from
+     * the reserve queue.
+     * @param index
+     */
+    private replaceActive(index: number) {
+        if (index >= 0) {
```

Comment:
NOP `if index < 0` also.

Reply:
same reason as for `done`... `gotError(removed_node_address)` will do a `NOP` here, but this is legit.